src/main/kotlin/gg/essential/elementa/components/Window.kt

```diff
@@ -357,13 +357,13 @@ class Window @JvmOverloads constructor(
             )
         }
 
-        fun ofOrNull(component: UIComponent): Window? {
+        fun ofOrNull(component: UIComponent): Window? = component.cachedWindow ?: run {
             var current = component
 
-            while (current !is Window && current.hasParent && current.parent != current)
+            while (current !is Window && current.hasParent && current.parent != current && current.cachedWindow == null)
                 current = current.parent
 
-            return current as? Window
+            (current as? Window ?: current.cachedWindow).also { component.cachedWindow = it }
```

Comment:
We must only store the cached value if the component is not currently hidden, i.e. only if `current in current.parent.children` is always true at the start of the loop body.

Reply:
Why exactly is this? The parent reference will mean the method returns the same as it did before for hidden components, and the cache will be invalidated when the component is unhidden anyway?

Reply:
Let's say we have window W and V and components A and B where W is parent of A is parent of B. Now, if we hide B, then call `Window.of(B)`, then W will be cached as B's window. But if we then change the parent of A to V, then `Window.of(B)` will still return W, even though it should return V.

Reply:
Ah, gotcha. What if we were to opt for a less cache-like approach in which we just recursively set the field of a new child to the current window when it is added, instead of only invalidating and relying on it being populated later?

Reply:
Oh, yeah, just setting it when it's added and resetting it when it's removed, that would indeed be simpler while having the exact same result.

Reply:
Ah yeah, I'll do that